Added testing chapter in Best Practices (#24)
ImperialCollegeLondon/ReCoDE-NavierStokesPropagator: docs/src/best_practices.md

`Simulation Util` instead, as discussed above, wraps around open-source libraries and algorithmic routines that aids in the simulation of the Navier-Stokes equations. `Simulation Constructs` and `Simulation Util` are divided in a way that source files in `Simulation Util` would only be dependent on `Simulation Constructs` but not vis versa.

## Testing

Here, we introduce the testing methodology for software, based on the `Test.jl` package in Julia's standard library. As software gets developed, modifications to the codebase would inevitably change the behaviour of outputs. By placing in safeguards in terms of automated testing (since it is software that can be run automatically by a microprocessor), bugs introduced during the development can be avoided and flagged to the developer in a manner appropriate. The automated testing of the codebase is part of the continuous integration (CI) software development practice, that we would touch upon below.

> In general, testing can be broken into levels of granularity. At the lowest level of granularity, we have unit tests, suitably written for testing the small chunks of code, generally functions and basis properties of structures. Climbing one level up, we have integration testing. Here, the focus is more on validation of correct software behaviour between inputs and outputs of functions. As software written gets complicated, systems level testing is introduced as the next level of granularity, which now is considered a black-box technique, usually tasked to be executed by software quality engineers instead of developers and hence we omit ourselves from the systems testing and higher level granularity at this point.

Reviewer: ```suggestion
In general, testing can be broken into levels of granularity. At the lowest level of granularity, we have unit tests, suitably written for testing the small chunks of code, generally functions and basic properties of structures. Climbing one level up, we have integration testing. Here, the focus is more on validation of correct software behaviour between inputs and outputs of functions. As the software written gets more complicated, systems level testing is introduced as the next level of granularity, which now is considered a black-box technique, usually tasked to be executed by software quality engineers instead of developers and hence we omit ourselves from the systems testing and higher level granularity at this point.
```

> In terms of unit test written, coverage has been written for `DomainDesriptors.jl`, `States.jl`, `LinearSolvers.jl` and `Transformations.jl`. Due to its nature of being a unit test, this package adopts the convention that the test files are named similarily to their respectively source files, and structure in a mirrored way to `src`. In Julia's `Test.jl` package, developers are able to assign test cases and apply assertions via the macro @test.

Reviewer: ```suggestion
In terms of the unit tests written, coverage has been written for `DomainDesriptors.jl`, `States.jl`, `LinearSolvers.jl` and `Transformations.jl`. Due to its nature of being a unit test, this package adopts the convention that the test files are named similarly to their respective source files, and structure in a mirrored way to `src`. In Julia's `Test.jl` package, developers are able to assign test cases and apply assertions via the `@test` macro.
```

```julia
@test <write statement here>
```
This is the core feature of testing, allowing for assertion to be made to ensure the correctness of the statement in the <> braces. Doing so in `test/sim/DomainDescriptors.jl`, we are able to test the constructor of `DomainDescriptor`, ensuring the features expected from the `DomainDescriptor` struct are adhered to.

> Furthermore, due to the multiple assertions being made in a test, the `@testset` macro allows developer to sort sets together, organising the test summary into digestable chunks. When organised wisely where each testset are sorted by common features, each testset can elucidate insight for the starting point of debugging the codebase when bugs are detected from testing.

Reviewer: ```suggestion
Furthermore, due to the multiple assertions being made in a test, the `@testset` macro allows the developer to sort sets together, organising the test summary into digestable chunks. When organised wisely where each testset are sorted by common features, each testset can elucidate insight for the starting point of debugging the codebase when bugs are detected from testing.
```

> Going one level up, the testing done by `StateTransformations.jl` starts integrating the properties of `State` which are to be transformed by properties in a `Transformer`. Since the interaction here are due to functions involving a `util` and a `sim` dependecy, it is considered as an integration test, and have resides in the `StateTransformations.jl` test file in `test/`.

Reviewer: ```suggestion
Going one level up, the testing done by `StateTransformations.jl` starts integrating the properties of `State` which are to be transformed by properties in a `Transformer`. Since the interaction here are due to functions involving a `util` and a `sim` dependency, it is considered as an integration test, and have resides in the `StateTransformations.jl` test file in `test/`.
```

> ### Unit Testing Code Coverage

Reviewer: ```suggestion
### Unit Testing Code Coverage

```

> Unit test should ideally be written for all cases, having coverage for all lines of the sourcecode. However, it is impractical as exhibit in this instance. The reader may notice that although unit testing have been written for a majority of the `Simulation Construct (\sim)` and `Simulation Utility (\util)` source files, none have been written for the `NavierStokesPropagators*.jl` source files. The reason is the invability of good quality test to be written as the simulation code itself are complicated and involved. We are dealing with the numerics of chaotic turbulence here! Hence, as a developer, it is important to realise the issue apriori and written safeguards into the codebase to guard from introducing bugs.

Reviewer: ```suggestion
Unit tests should ideally be written for all cases, having coverage for all lines of the source code. However, it is impractical as exhibit in this instance. The reader may notice that although the unit tests have been written for most of the `Simulation Construct (\sim)` and `Simulation Utility (\util)` source files, none have been written for the `NavierStokesPropagators*.jl` source files. The reason is the unavailability of good quality tests to be written as the simulation code itself are complicated and involved. We are dealing with the numerics of chaotic turbulence here! Hence, as a developer, it is important to realise the issue apriori and include safeguards into the codebase to guard from introducing bugs.
```

> For this simulation, the state arrays are can be in `frac_mode` and `freq_mode` (essentiall the frequency domain). However, the hundreds of lines of code obfuscate the currently state (`frac_mode` and `freq_mode`) when a function in `NavierStokesPropagators*.jl` runs. As a result, checks are place as safeguards to the entry of a function, for example,

Reviewer: ```suggestion
For this simulation, the state arrays are be in the `frac_mode` and `freq_mode` (essentially in the frequency domain). However, the hundreds of lines of code obfuscate the currently state (`frac_mode` and `freq_mode`) when a function in `NavierStokesPropagators*.jl` runs. As a result, checks are placed as safeguards to the entry of a function, for example,
```
